src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs

```diff
@@ -1453,4 +1454,99 @@ public void ReportDiagnostic(Diagnostic diagnostic)
+        /// </summary>
+        public ImmutableArray<Diagnostic> ReportedDiagnostics { get; }
+
+        /// <summary>
+        /// <see cref="CodeAnalysis.Compilation"/> for the context.
+        /// </summary>
+        public Compilation Compilation { get; }
+
+        /// <summary>
+        /// Options specified for the analysis.
+        /// </summary>
+        public AnalyzerOptions Options { get; }
+
+        /// <summary>
+        /// Token to check for requested cancellation of the analysis.
+        /// </summary>
+        public CancellationToken CancellationToken { get; }
+
+        internal SuppressionAnalysisContext(
+            Compilation compilation,
+            AnalyzerOptions options,
+            ImmutableArray<Diagnostic> reportedDiagnostics,
+            Action<Suppression> suppressDiagnostic,
+            Func<SuppressionDescriptor, bool> isSupportedSuppressionDescriptor,
+            Func<SyntaxTree, SemanticModel> getSemanticModel,
+            CancellationToken cancellationToken)
+        {
+            Compilation = compilation;
+            Options = options;
+            ReportedDiagnostics = reportedDiagnostics;
+            _addSuppression = suppressDiagnostic;
+            _isSupportedSuppressionDescriptor = isSupportedSuppressionDescriptor;
+            _getSemanticModel = getSemanticModel;
+            CancellationToken = cancellationToken;
+        }
+
+        /// <summary>
+        /// Report a <see cref="Suppression"/> for a reported diagnostic.
+        /// </summary>
+        public void ReportSuppression(Suppression suppression)
+        {
+            if (!ReportedDiagnostics.Contains(suppression.SuppressedDiagnostic))
+            {
+                // Non-reported diagnostic with ID '{0}' cannot be suppressed.
+                var message = string.Format(CodeAnalysisResources.NonReportedDiagnosticCannotBeSuppressed, suppression.SuppressedDiagnostic.Id);
+                throw new ArgumentException(message);
+            }
+
+            if (!_isSupportedSuppressionDescriptor(suppression.Descriptor))
+            {
+                // Reported suppression with ID '{0}' is not supported by the suppressor.
+                var message = string.Format(CodeAnalysisResources.UnsupportedSuppressionReported, suppression.Descriptor.Id);
+                throw new ArgumentException(message);
+            }
+
+            if (!Compilation.SyntaxTrees.Any(t => t.Options.Features.ContainsKey(AnalyzerDriver.DiagnosticSuppressorFeatureName)))
+            {
+                throw new ArgumentException(CodeAnalysisResources.DiagnosticSuppressorFeatureDisabled);
+            }
+
+            if (suppression.Descriptor.IsDisabled(Compilation.Options))
+            {
+                // Suppression has been disabled by the end user through compilation options.
+                return;
+            }
+
+            lock (_addSuppression)
```

Comment:
>lock (_addSuppression) [](start = 12, length = 22)

Instead of locking on the delegate, we probably move synchronization into the delegate. The function should have more idea whether a lock is needed and on what.

Reply:
I have followed the implementation used by the existing context implementations. We can possibly clean this up for the whole file based on performance measurements. Tagging @sharwell

Reply:
> I have followed the implementation used by the existing context implementations. We can possibly clean this up for the whole file based on performance measurements. Tagging @sharwell 

locking on a delegate instance feels wrong to me. Even if other code does similar things, I see no good reason to do this just because other parts do the same wrong thing. Could you point me to all possible delegates that can end up here?

---
In reply to: [292216857](https://github.com/dotnet/roslyn/pull/36067#discussion_r292216857) [](ancestors = 292216857)

Reply:
The delegate passed in (AnalyzerDriver.AddSuppression) is already thread-safe, so I am going to just remove the lock here.

Reply:
> The delegate passed in (AnalyzerDriver.AddSuppression) is already thread-safe, so I am going to just remove the lock here.

Consider adjusting doc comments where appropriate to mention that the delegate should be thread-safe.

---
In reply to: [292578643](https://github.com/dotnet/roslyn/pull/36067#discussion_r292578643) [](ancestors = 292578643)

Reply:
Thanks, done.